### ISSUE TITLE
limit and  last queries consuming > 15 GB memory

### ISSUE THREAD

**Issue by joshuajoh** [influxdata/influxdb#7587]
### Bug report

__System info:__  InfluxDB 1.0.0 and 1.0.2, Red Hat Enterprise Linux Server release 6.8

__Steps to reproduce:__

1. 40 servers running telegraf with procstat to capture all process info every 1 minute
2. target database with one hour retention period
3. average number of series per server is 550 in the database, but one server has 16K
4. execute one of the following queries:

    select * from  procstat limit 5

    SELECT last("pid") as pid, last("cpu_usage") as cpu, last("memory_rss") / 1048576  as memory, last("num_fds") as fds, last("num_threads") as threads FROM "procstat" WHERE "host" =~ /^Server_With_16K_Series$/ AND time > now() - 2m GROUP BY "process_name" fill(null)

    a.  The limit query will cause 15 GB to be consumed instantly
    b.  The last query running every 2 minutes will consume all available memory.  In Grafana the query returns, but SHOW QUERIES will show instances still running:

> show queries
qid     query                                                                                                                                                                                                                                                                 database         duration
612     SELECT last(pid) AS pid, last(cpu_usage) AS cpu, last(memory_rss) / 1048576 AS memory_in_MB, last(num_fds) AS fds, last(num_threads) AS threads, last(cpu_time_iowait) AS cpu_iowait, last(read_count) AS read_count, last(write_count) AS write_count FROM procstat WHERE host =~ /^Server_With_16K_Series$/ AND time > now() - 2m GROUP BY process_name  telegraf_procstat       13m10s
246     SELECT last(pid) AS pid, last(cpu_usage) AS cpu, last(memory_rss) / 1048576 AS memory_in_MB, last(num_fds) AS fds, last(num_threads) AS threads, last(cpu_time_iowait) AS cpu_iowait, last(read_count) AS read_count, last(write_count) AS write_count FROM procstat WHERE host =~ /^Server_With_16K_Series$/ AND time > now() - 2m GROUP BY process_name  telegraf_procstat       23m11s
791     SELECT last(pid) AS pid, last(cpu_usage) AS cpu, last(memory_rss) / 1048576 AS memory_in_MB, last(num_fds) AS fds, last(num_threads) AS threads, last(cpu_time_iowait) AS cpu_iowait, last(read_count) AS read_count, last(write_count) AS write_count FROM procstat WHERE host =~ /^Server_With_16K_Series$/ AND time > now() - 2m GROUP BY process_name  telegraf_procstat       3m10s
27      SELECT last(pid) AS pid, last(cpu_usage) AS cpu, last(memory_rss) / 1048576 AS memory_in_MB, last(num_fds) AS fds, last(num_threads) AS threads, last(cpu_time_iowait) AS cpu_iowait, last(read_count) AS read_count, last(write_count) AS write_count FROM procstat WHERE host =~ /^Server_With_16K_Series$/ AND time > now() - 2m GROUP BY process_name  telegraf_procstat       28m35s
32      SELECT last(pid) AS pid, last(cpu_usage) AS cpu, last(memory_rss) / 1048576 AS memory_in_MB, last(num_fds) AS fds, last(num_threads) AS threads, last(cpu_time_iowait) AS cpu_iowait, last(read_count) AS read_count, last(write_count) AS write_count FROM procstat WHERE host =~ /^Server_With_16K_Series$/ AND time > now() - 2m GROUP BY process_name  telegraf_procstat       27m11s
133     SELECT last(pid) AS pid, last(cpu_usage) AS cpu, last(memory_rss) / 1048576 AS memory_in_MB, last(num_fds) AS fds, last(num_threads) AS threads, last(cpu_time_iowait) AS cpu_iowait, last(read_count) AS read_count, last(write_count) AS write_count FROM procstat WHERE host =~ /^Server_With_16K_Series$/ AND time > now() - 2m GROUP BY process_name  telegraf_procstat       25m11s
455     SELECT last(pid) AS pid, last(cpu_usage) AS cpu, last(memory_rss) / 1048576 AS memory_in_MB, last(num_fds) AS fds, last(num_threads) AS threads, last(cpu_time_iowait) AS cpu_iowait, last(read_count) AS read_count, last(write_count) AS write_count FROM procstat WHERE host =~ /^Server_With_16K_Series$/ AND time > now() - 2m GROUP BY process_name  telegraf_procstat       19m10s
552     SELECT last(pid) AS pid, last(cpu_usage) AS cpu, last(memory_rss) / 1048576 AS memory_in_MB, last(num_fds) AS fds, last(num_threads) AS threads, last(cpu_time_iowait) AS cpu_iowait, last(read_count) AS read_count, last(write_count) AS write_count FROM procstat WHERE host =~ /^Server_With_16K_Series$/ AND time > now() - 2m GROUP BY process_name  telegraf_procstat       17m10s
380     SELECT last(pid) AS pid, last(cpu_usage) AS cpu, last(memory_rss) / 1048576 AS memory_in_MB, last(num_fds) AS fds, last(num_threads) AS threads, last(cpu_time_iowait) AS cpu_iowait, last(read_count) AS read_count, last(write_count) AS write_count FROM procstat WHERE host =~ /^Server_With_16K_Series$/ AND time > now() - 2m GROUP BY process_name  telegraf_procstat       21m10s
714     SELECT last(pid) AS pid, last(cpu_usage) AS cpu, last(memory_rss) / 1048576 AS memory_in_MB, last(num_fds) AS fds, last(num_threads) AS threads, last(cpu_time_iowait) AS cpu_iowait, last(read_count) AS read_count, last(write_count) AS write_count FROM procstat WHERE host =~ /^Server_With_16K_Series$/ AND time > now() - 2m GROUP BY process_name  telegraf_procstat       7m10s
643     SELECT last(pid) AS pid, last(cpu_usage) AS cpu, last(memory_rss) / 1048576 AS memory_in_MB, last(num_fds) AS fds, last(num_threads) AS threads, last(cpu_time_iowait) AS cpu_iowait, last(read_count) AS read_count, last(write_count) AS write_count FROM procstat WHERE host =~ /^Server_With_16K_Series$/ AND time > now() - 2m GROUP BY process_name  telegraf_procstat       11m10s
672     SELECT last(pid) AS pid, last(cpu_usage) AS cpu, last(memory_rss) / 1048576 AS memory_in_MB, last(num_fds) AS fds, last(num_threads) AS threads, last(cpu_time_iowait) AS cpu_iowait, last(read_count) AS read_count, last(write_count) AS write_count FROM procstat WHERE host =~ /^Server_With_16K_Series$/ AND time > now() - 2m GROUP BY process_name  telegraf_procstat       9m10s
748     SELECT last(pid) AS pid, last(cpu_usage) AS cpu, last(memory_rss) / 1048576 AS memory_in_MB, last(num_fds) AS fds, last(num_threads) AS threads, last(cpu_time_iowait) AS cpu_iowait, last(read_count) AS read_count, last(write_count) AS write_count FROM procstat WHERE host =~ /^Server_With_16K_Series$/ AND time > now() - 2m GROUP BY process_name  telegraf_procstat       5m10s
582     SELECT last(pid) AS pid, last(cpu_usage) AS cpu, last(memory_rss) / 1048576 AS memory_in_MB, last(num_fds) AS fds, last(num_threads) AS threads, last(cpu_time_iowait) AS cpu_iowait, last(read_count) AS read_count, last(write_count) AS write_count FROM procstat WHERE host =~ /^Server_With_16K_Series$/ AND time > now() - 2m GROUP BY process_name  telegraf_procstat       15m10s
857     SELECT last(pid) AS pid, last(cpu_usage) AS cpu, last(memory_rss) / 1048576 AS memory_in_MB, last(num_fds) AS fds, last(num_threads) AS threads, last(cpu_time_iowait) AS cpu_iowait, last(read_count) AS read_count, last(write_count) AS write_count FROM procstat WHERE host =~ /^Server_With_16K_Series$/ AND time > now() - 2m GROUP BY process_name  telegraf_procstat       1m10s


__Expected behavior:__ LIMIT with small value or grabbing the last two minutes of a large series should not consume large amount of memory

__Actual behavior:__ 1) A database of 300 MB consumes more than 15 GB of RAM when using LIMIT without WHERE clause 2) all available memory is consumed when querying the last two minutes of data periodically and using LAST 

__Additional info:__ Totally expect a performance hit for a large series SELECT, but this should be counteracted by the use of a 2m WHERE clause and LAST.  LIMIT with a small value should always be a safe query.

__Please let me know if any other informaion is required.__ 

Thank-you.

**Comment by jwilder** [comment]
Can you run the following when you see the higher memory usage and attach the output in a gist? 

```
curl -o block.txt "http://localhost:8086/debug/pprof/block?debug=1" 
curl -o goroutine.txt "http://localhost:8086/debug/pprof/goroutine?debug=1" 
curl -o heap.txt "http://localhost:8086/debug/pprof/heap?debug=1" 
curl -o vars.txt "http://localhost:8086/debug/vars" 
iostat -xd 1 30 > iostat.txt
influx -execute "show shards" > shards.txt
influx -execute "show stats" > stats.txt
influx -execute "show diagnostics" > diagnostics.txt
```

Also, can you try `1.1rc1` and see if you still have the same issue?

**Comment by joshuajoh** [comment]
Hi @jwilder,

I was able to consume all the memory during the second execution of:

`select * from procstat limit 5`

Iostat may not be accurate as it was taken during swap thrashing, let me know if you need another run of it or the gist format is no good.

https://gist.github.com/joshuajoh/ea0281d1b952886749286a24ddd844e9

I'll give 1.1rc1 a shot, I'm also going to play with rewriting the last query to something like this too:






SELECT pid as pid, cpu_usage as cpu, "memory_rss" / 1048576 as memory, "num_fds" as fds, "num_threads" as threads FROM "procstat" WHERE "host" =~ /^Server_With_16K_Series$/ AND time > now() - 90s GROUP BY "process_name" ORDER BY desc LIMIT 1

Thank-you.

**Comment by ehartmann** [comment]
Hi,

I think we are facing the same issue (InfluxDB 1.0.2 CentOS 7). 
A simple query like :
`select * from "apache.hit" where host='server15.ec2.sonarsource.io' order by time desc limit 5` 
is making InfluxDB crashing. Here are an excerpt of /var/log/messages : 
`Nov  8 17:24:03 server10 influxd: goroutine 59979 [IO wait]:
Nov  8 17:24:03 server10 influxd: net.runtime_pollWait(0x7f780a142600, 0x72, 0xc9ce340000)
Nov  8 17:24:03 server10 influxd: /usr/local/go/src/runtime/netpoll.go:160 +0x60
Nov  8 17:24:03 server10 influxd: net.(*pollDesc).Wait(0xc92743ea70, 0x72, 0x0, 0x0)
Nov  8 17:24:03 server10 influxd: /usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
Nov  8 17:24:03 server10 influxd: net.(*pollDesc).WaitRead(0xc92743ea70, 0x0, 0x0)
Nov  8 17:24:03 server10 influxd: /usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
Nov  8 17:24:03 server10 influxd: net.(*netFD).Read(0xc92743ea10, 0xc9ce340000, 0x1000, 0x1000, 0x0, 0x7f7c443c4050, 0xc82000e150)
Nov  8 17:24:03 server10 influxd: /usr/local/go/src/net/fd_unix.go:250 +0x23a
Nov  8 17:24:03 server10 influxd: net.(*conn).Read(0xc9c9621728, 0xc9ce340000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
Nov  8 17:24:03 server10 influxd: /usr/local/go/src/net/net.go:172 +0xe4
Nov  8 17:24:03 server10 influxd: net/http.noteEOFReader.Read(0x7f7c28593e50, 0xc9c9621728, 0xc9c5ad53e8, 0xc9ce340000, 0x1000, 0x1000, 0x60, 0x0, 0x0)
Nov  8 17:24:03 server10 influxd: /usr/local/go/src/net/http/transport.go:1687 +0x67
Nov  8 17:24:03 server10 influxd: net/http.(*noteEOFReader).Read(0xc9ce332b60, 0xc9ce340000, 0x1000, 0x1000, 0xc8abf26f98, 0x0, 0x0)
Nov  8 17:24:03 server10 influxd: <autogenerated>:284 +0xd0
Nov  8 17:24:03 server10 influxd: bufio.(*Reader).fill(0xc8eb9330e0)
Nov  8 17:24:03 server10 influxd: /usr/local/go/src/bufio/bufio.go:97 +0x1e9
Nov  8 17:24:03 server10 influxd: bufio.(*Reader).Peek(0xc8eb9330e0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
Nov  8 17:24:03 server10 influxd: /usr/local/go/src/bufio/bufio.go:132 +0xcc
Nov  8 17:24:03 server10 influxd: net/http.(*persistConn).readLoop(0xc9c5ad5380)
Nov  8 17:24:03 server10 influxd: /usr/local/go/src/net/http/transport.go:1073 +0x177
Nov  8 17:24:03 server10 influxd: created by net/http.(*Transport).dialConn
Nov  8 17:24:03 server10 influxd: /usr/local/go/src/net/http/transport.go:857 +0x10a6
Nov  8 17:24:03 server10 influxd: goroutine 58226 [IO wait]:
Nov  8 17:24:03 server10 influxd: net.runtime_pollWait(0x7f7c28595128, 0x72, 0xc8bb2ee000)
Nov  8 17:24:03 server10 influxd: /usr/local/go/src/runtime/netpoll.go:160 +0x60
Nov  8 17:24:03 server10 influxd: net.(*pollDesc).Wait(0xc93a657410, 0x72, 0x0, 0x0)
Nov  8 17:24:03 server10 influxd: /usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
Nov  8 17:24:03 server10 influxd: net.(*pollDesc).WaitRead(0xc93a657410, 0x0, 0x0)
Nov  8 17:24:03 server10 influxd: /usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
Nov  8 17:24:03 server10 influxd: net.(*netFD).Read(0xc93a6573b0, 0xc8bb2ee000, 0x1000, 0x1000, 0x0, 0x7f7c443c4050, 0xc82000e150)
Nov  8 17:24:03 server10 influxd: /usr/local/go/src/net/fd_unix.go:250 +0x23a
Nov  8 17:24:03 server10 influxd: net.(*conn).Read(0xc96f66f448, 0xc8bb2ee000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
Nov  8 17:24:03 server10 influxd: /usr/local/go/src/net/net.go:172 +0xe4
Nov  8 17:24:03 server10 influxd: net/http.(*connReader).Read(0xc8e4c59ce0, 0xc8bb2ee000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
Nov  8 17:24:03 server10 influxd: /usr/local/go/src/net/http/server.go:526 +0x196
Nov  8 17:24:03 server10 influxd: bufio.(*Reader).fill(0xc8bb2ec240)
Nov  8 17:24:03 server10 influxd: /usr/local/go/src/bufio/bufio.go:97 +0x1e9
Nov  8 17:24:03 server10 influxd: bufio.(*Reader).Peek(0xc8bb2ec240, 0x4, 0x0, 0x0, 0x0, 0x0, 0x0)
Nov  8 17:24:03 server10 influxd: /usr/local/go/src/bufio/bufio.go:132 +0xcc
Nov  8 17:24:03 server10 influxd: net/http.(*conn).readRequest(0xc97ff12700, 0x0, 0x0, 0x0)
Nov  8 17:24:03 server10 influxd: /usr/local/go/src/net/http/server.go:702 +0x2e6
Nov  8 17:24:03 server10 influxd: net/http.(*conn).serve(0xc97ff12700)
Nov  8 17:24:03 server10 influxd: /usr/local/go/src/net/http/server.go:1425 +0x947
Nov  8 17:24:03 server10 influxd: created by net/http.(*Server).Serve
Nov  8 17:24:03 server10 influxd: /usr/local/go/src/net/http/server.go:2137 +0x44e
Nov  8 17:24:04 server10 systemd: influxdb.service: main process exited, code=exited, status=2/INVALIDARGUMENT
Nov  8 17:24:04 server10 systemd: Unit influxdb.service entered failed state.
Nov  8 17:24:04 server10 systemd: influxdb.service failed.
Nov  8 17:24:04 server10 systemd: influxdb.service holdoff time over, scheduling restart.
`
@jwilder, I can make a try upgrading to 1.1.0-rc1 but is it possible to rollback from 1.1.0-rc1 to 1.0.2 ?

**Comment by joshuajoh** [comment]
So the default Grafana editor automatically makes every template variable a regexp for InfluxDB.

As soon as the queries were updated to the below memory consumption is fine.

`WHERE host = 'Server_With_16K_Series' AND time > now() - 2m GROUP BY process_name`

Looks like there is some type of memory consumption issue when using a regular expression for 40K series.

Please let me know if there is any other information that can be provided for this.

Thank-you.

**Comment by jwilder** [comment]
Can you show the slower regex version of the query?

**Comment by joshuajoh** [comment]
Sure:

`SELECT last(pid) AS pid, last(cpu_usage) AS cpu, last(memory_rss) / 1048576 AS memory_in_MB, last(num_fds) AS fds, last(num_threads) AS threads, last(cpu_time_iowait) AS cpu_iowait, last(read_count) AS read_count, last(write_count) AS write_count FROM procstat WHERE host =~ /^Server_With_16K_Series$/ AND time > now() - 2m GROUP BY process_name`

**Comment by jwilder** [comment]
That specific regex is automatically rewritten to use equality in 1.1. I'd suggest testing 1.1rc1.

**Comment by jwilder** [comment]
Should be fixed in 1.1.
